internal/jimm/model_test.go

```diff
@@ -1054,6 +1054,105 @@ users:
 		OwnerTag: names.NewUserTag("alice@canonical.com").String(),
 	},
 	expectError: "no cloud specified for model; please specify one",
+}, {
+	name: "CreateModelOnACloudWithNoRegions",
+	env: `
+clouds:
+- name: test-cloud
+  type: test-provider
+  regions:
+  - name: default
+    virtual: true  // this is a virtual cloud region
+  users:
+  - user: alice@canonical.com
+    access: add-model
+cloud-credentials:
+- name: test-credential-1
+  owner: alice@canonical.com
+  cloud: test-cloud
+  auth-type: empty
+controllers:
+- name: controller-1
+  uuid: 00000000-0000-0000-0000-0000-0000000000001
+  cloud: test-cloud
+  region: default
+  cloud-regions:
+  - cloud: test-cloud
+    region: default
+    priority: 0
+- name: controller-2
+  uuid: 00000000-0000-0000-0000-0000-0000000000002
+  cloud: test-cloud
+  region: default
+  cloud-regions:
+  - cloud: test-cloud
+    region: default
+    priority: 2
+`[1:],
+	updateCredential: func(_ context.Context, _ jujuparams.TaggedCredential) ([]jujuparams.UpdateCredentialModelResult, error) {
+		return nil, nil
+	},
+	grantJIMMModelAdmin: func(_ context.Context, _ names.ModelTag) error {
+		return nil
+	},
+	createModel: assertCreateModelArgs(&jujuparams.ModelCreateArgs{
```

Comment:
This test case and the corresponding `assertCreateModelArgs` are super confusing to read. Why is this pattern needed?

Reply:
just continuing the pattern that was here before which is not all that confusing.. think of it as middleware: each function checks a certain aspect of the call and passes check to the next one

Reply:
`expectModel` asserts that our database looks the way we expect it - it does not assert the arguments sent to the controller to create a new model. 
the `assertCreateModelArgs` asserts that what we send to the controller contains parameters we thing should be sent to the controller - in this test case the cloud region should not be sent since it's a virtual one

Reply:
Ahh I understand, that is nice but tough to parse. Grabbing lunch then I'll take a closer look with that in mind.